### ISSUE TITLE
Non-partial models which are not locally balanced should generate an error in the NF when checking and trying to simulate

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#8024]
Consider the following MWE
```
model M
  Real x, y;
equation
  x = 3;
end M;
```
If I check the model, I get 
```
[1] 11:31:10 Scripting Notification
Check of M completed successfully.
```
In fact, this model is not partial, so it should be balanced. If it isn't, that is for sure an error. This is currently reported by the backend later on when trying to simulate
```
[1] 11:32:11 Symbolic Error
Too few equations, under-determined system. The model has 1 equation(s) and 2 variable(s).
```
but in fact, this information is already available to the NF, which should report it.

Proposed output when checking:
```
[1] 11:31:10   Symbolic Error
The model M is not balanced.
```

**Comment by casella** [comment]
> Sorry, I missed that you were already doing check model. In that case you get the message:
> 
> ```
> Check of M completed successfully.
> Class M has 1 equation(s) and 2 variable(s).
> 1 of these are trivial equation(s).
> ```

My point is that, since the model is not partial, in this case the outcome should _not_ be "Check of M completed successfully" but rather "Model M is not balanced". Which is an error, not a notification.

**Comment by adeas31** [comment]
The simplest fix is to remove the message "Check of M completed successfully" from the output of `checkModel` and simply show the rest of the result i.e., 

```
class M has 1 equation(s) and 2 variable(s).
1 of these are trivial equation(s).
```

**Comment by casella** [comment]
> Can you just carry around a flag during flattening, and make it true when you encounter a partial class (instead of generating an error and quit)?

@perost, I'm trying to figure out if we can get through this ticket or if we should just give up until we implement proper balancing checks in the future.

Would my last proposal be feasible, or is it too much work? I think understanding whether a certain class contains partial stuff or not could be valuable; in fact, we already do that during normal flattening (and abort in that case) and we're simply suppressing this check during CheckModel precisely because a model being checked doesn't necessarily need to be runnable.

What do you think?

**Comment by casella** [comment]
@j-emils I agree with you in principle, if we had the proper balancing analysis implemented, as explained in [Section 4.7](https://specification.modelica.org/maint/3.6/class-predefined-types-and-declarations.html#balanced-models) of the MSL.

As mentioned in [my previous comment](https://github.com/OpenModelica/OpenModelica/issues/8024#issuecomment-949743200), having the same number of variables and equations is not necessary for a (component) model to be valid. A reusable component model stored in a package that has some replaceable class or component which is set to a partial class is valid per se, though of course it cannot be used directly, it needs those replaceable things to be redeclared. But it would be quite bad if Check Model failed on it.

See #3977. We'll need to work on that at some point.

**Comment by j-emils** [comment]
@casella I understand. However, the model you're describing I would call a partial model that should be declared with the partial keyword. In that case the check should not fail.

**Comment by j-emils** [comment]
I understand, and of course you're right in a sense. However, I guess we're getting into a discussion about semantics, and particularly the word "incomplete". Looking at the description of how the partial keyword should be used:

"A class defined with partial in the class-prefixes is called a partial class. Such a class is allowed to be incomplete, and cannot be instantiated in a simulation model; useful, e.g., as a base class." 

I would say that a model that is not globally balanced would be incomplete, since it's not executable. But maybe one should distinguish between complete and executable?

Either way, I agree with you that the diagnostics is far better when displaying what is actually wrong or missing.

**Comment by casella** [comment]
Well, non partial Modelica.Fluid models can indeed be instantiated. They just need the Medium model to be redeclared from a partial one to a concrete one. If your definition of "complete" is "instantiable", everything fits.

In any case, the Modelica Specification is not the Gospel, and it could of course be changed. But that's out of scope here, the goal of OSMC is to implement the standard in force. If you have suggestions to change the Modelica semantics, the right venue for that discussion is https://github.com/modelica/ModelicaSpecification/issues. Good luck with changing the Specification in a way that would make existing valid models invalid 😃
